### ISSUE TITLE
Lighting support

### ISSUE THREAD

**Issue by caelunshun** [caelunshun/feather#60]

**Comment by NyxCode** [comment]
Just stumbled over this project, looks awesome! I thought i'd try to play around with this a bit to learn a bit about spec.

What about a `System` which goes through all `BlockUpdateEvent`s and recalculates lighting in every chunk which has been updated? Also, where would the light level of a block be stored? In its own `Component`?

**Comment by caelunshun** [comment]
> Just stumbled over this project, looks awesome! I thought i'd try to play around with this a bit to learn about spec.
> 
> What about a `System` which goes through all `BlockUpdateEvent`s and recalculates lighting in every chunk which has been updated?

Lighting would be stored in the `Chunk`, as it is in vanilla, in two compact arrays. See https://wiki.vg/index.php?title=Chunk_Format&diff=14678&oldid=14646#Chunk_Section_structure.

As a starting point, going through `BlockUpdateEvent`s and recalculating the entire chunk would be fine. We would want to optimize the algorithm so that only blocks affected by light updates are recalculated, and not the entire chunk.

**Comment by NyxCode** [comment]
that sounds reasonable. 
I think a good approach would be to store all light sources in each chunk. When a light source is placed/destroyed, we just increase/decrease the light-level of surrounding blocks. When a normal block is placed/broken though, we'll have to go through all light sources (which are close) and recalculate their light. What do you think?
I think for that it'd be nice to create a struct similar to `ChunkMap` which just stores the light sources, soemthing like `HashMap<ChunkPosition, Vec<(BlockPosition, LightLevel)>>.

**Comment by caelunshun** [comment]
@NyxCode Yes, that will work. To start, we should probably resolve #103 so that existing lighting data is loaded correctly. This will also involve adding the light storage in `Chunk`.

**Comment by NyxCode** [comment]
yeah, i'll look at that first! I don't have a lot of time, but this project looks great! 
Is there a place for discussions/questions except github issues? maybe a discord server?

**Comment by caelunshun** [comment]
> Is there a place for discussions/questions except github issues? maybe a discord server?
Well, the project didn't have any publicity until it ballooned yesterday, so we don't really have a community or discussion place. I could make a Discord server, but given the early state of the project, I'm not sure if there's really any point.

**Comment by NyxCode** [comment]
:D the post on the rust subreddit, huh? 
I mean, it's your call, but I could imagine that a place for discussion on the development would be usefull.

**Comment by caelunshun** [comment]
> I mean, it's your call, but I could imagine that a place for discussion on the development would be usefull.

I guess I agree; here's the Discord server. https://discord.gg/4eYmK69
